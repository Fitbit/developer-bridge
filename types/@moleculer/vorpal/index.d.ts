// Typedoc definitions taken and modified from vorpal source https://github.com/dthree/vorpal

/// <reference types="inquirer" />
declare module '@moleculer/vorpal' {
  import { EventEmitter } from 'events';
  import { Inquirer } from 'inquirer';

  class Vorpal {
    activeCommand: Vorpal.CommandInstance;
    session: Vorpal.Session;
    ui: Vorpal.UiInstance;

    /**
     * Set id for command line history
     * @param id
     * @return {Vorpal}
     * @api public
     */
    history(id: string): this;

    /**
     * Imports a library of Vorpal API commands
     * from another Node module as an extension
     * of Vorpal.
     *
     * @param {Function} commands
     * @return {Vorpal}
     * @api public
     */
    use(commands: (this: this, vorpal: this) => any): this;

    /**
     * Registers a new command in the vorpal API.
     *
     * @param {String} name
     * @param {String} desc
     * @return {Command}
     * @api public
     */
    command(name: string, desc?: string): Vorpal.Command;

    /**
     * Sets the permanent delimiter for this
     * Vorpal server instance.
     *
     * @param {String} str
     * @return {Vorpal}
     * @api public
     */
    delimiter(str: string): this;

    /**
     * Hook the tty prompt to this given instance
     * of vorpal.
     *
     * @return {Vorpal}
     * @api public
     */
    show(): this;

    /**
     * Executes a vorpal API command and
     * returns the response either through a
     * callback or Promise in the absence
     * of a callback.
     *
     * A little black magic here - because
     * we sometimes have to send commands 10
     * miles upstream through 80 other instances
     * of vorpal and we aren't going to send
     * the callback / promise with us on that
     * trip, we store the command, callback,
     * resolve and reject objects (as they apply)
     * in a local vorpal._command variable.
     *
     * When the command eventually comes back
     * downstream, we dig up the callbacks and
     * finally resolve or reject the promise, etc.
     *
     * Lastly, to add some more complexity, we throw
     * command and callbacks into a queue that will
     * be unearthed and sent in due time.
     *
     * @param {String} cmd
     * @param {Object} args
     * @return {Promise}
     * @api public
     */
    exec<T>(
      cmd: string,
      args: { [k: string]: any; sessionId?: string },
    ): Promise<T>;
    exec<T>(cmd: string): Promise<T>;

    /**
     * Executes a Vorpal command in sync.
     *
     * @param {String} cmd
     * @param {Object} args
     * @return {*} stdout
     * @api public
     */
    execSync<T>(cmd: string): T;

    /**
     * Delegates to ui.log.
     *
     * @param {String} log
     * @return {Vorpal}
     * @api public
     */
    log(message?: any, ...optionalParams: any[]): this;

    /**
     * Registers a new 'mode' command in the vorpal API.
     *
     * @param {String} name
     * @param {String} desc
     * @return {Command}
     * @api public
     */
    mode(name: string, desc?: string): Vorpal.Command;
  }

  namespace Vorpal {
    interface Command {
      /**
       * Defines an action for a given command.
       *
       * @param {Function} fn
       * @return {Command}
       * @api public
       */
      action<T>(
        actionFn: (this: CommandInstance, args: Args) => void | Promise<T>,
      ): this;
      action<T>(
        actionFn: (this: CommandInstance, command: string) => void | Promise<T>,
      ): this;

      /**
       * Provides an alias to the command
       *
       * @param {String} alias
       * @return {Command}
       * @api public
       */
      alias(alias: string): this;

      /**
       * Sets args for static typing of options
       * using minimist.
       *
       * @param {Object} types
       * @return {Command}
       * @api public
       */
      types(types: { string: string[] }): this;

      /**
       * Doesn't show command in the help menu.
       *
       * @return {Command}
       * @api public
       */
      hidden(): this;

      /**
       * Defines description for given command.
       *
       * @param {String} str
       * @return {Command}
       * @api public
       */
      description(str: string): this;

      /**
       * Defines a prompt delimiter for a
       * mode once entered.
       *
       * @param {String} delimiter
       * @return {Command}
       * @api public
       */
      delimiter(delimiter: string): this;

      /**
       * Defines an init action for a mode command.
       *
       * @param {Function} fn
       * @return {Command}
       * @api public
       */
      init(fn: () => void | Promise<void>): this;

      /**
       * Registers an option for given command.
       *
       * @param {String} flags
       * @param {String} description
       * @return {Command}
       * @api public
       */
      option(
        flags: string,
        description: string,
        autocomplete?: Autocomplete,
      ): this;

      /**
       * Defines a function to be called when the
       * command is canceled.
       *
       * @param fn
       * @returns {Command}
       * @api public
       */

      cancel(fn: () => void): this;
    }

    type Autocomplete =
      | string[]
      | ((input: string) => string[] | Promise<string>)
      | ((input: string, callback: (completions: string[]) => void) => void);

    interface Args {
      options: {
        [name: string]: string;
      };
    }

    interface CommandInstance {
      log(message?: any, ...optionalParams: any[]): void;
      prompt: Inquirer['prompt'];
    }

    interface Session {
      id: number;
    }

    interface Redraw {
      (...message: string[]): void;
      done(): void;
    }

    interface UiInstance extends EventEmitter {
      redraw: Redraw;
      log(message?: any, ...optionalParams: any[]): void;
    }
  }

  export = Vorpal;
}
