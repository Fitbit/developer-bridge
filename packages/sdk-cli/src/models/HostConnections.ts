import 'stream.finished/auto';

import { RemoteHost } from '@fitbit/fdb-debugger';
import dateformat from 'dateformat';
import fs from 'fs';
import stream, { Duplex } from 'stream';
import { SyncEvent } from 'ts-events';

import * as developerRelay from '../api/developerRelay';
import * as USBDebugHost from './USBDebugHost';
import { DeviceType, HostType } from './HostTypes';

import StreamTap from './StreamTap';

export class HostConnection {
  private constructor(public stream: stream.Duplex, public host: RemoteHost) {}

  static getDumpStreamTap() {
    const shouldDumpLogFile = process.env.FITBIT_DEVBRIDGE_DUMP === '1';
    if (!shouldDumpLogFile) return undefined;

    const dumpLogFilePath = dateformat('"log" yyyy-mm-dd "at" H.MM.ss."txt"');
    const dumpLogFileHandle = fs.openSync(dumpLogFilePath, 'w');
    const now = () => new Date().getTime();
    const epoch = now();

    function writeChunk(prefix: string) {
      return (chunk: any) =>
        fs.writeSync(
          dumpLogFileHandle,
          `[${prefix}][${now() - epoch}] ${JSON.stringify(
            chunk,
            undefined,
            2,
          )}\n`,
        );
    }

    const transforms = {
      preSerializeTransform: new StreamTap(writeChunk('send')),
      postDeserializeTransform: new StreamTap(writeChunk('recv')),
    };

    stream.finished(transforms.postDeserializeTransform, () =>
      fs.closeSync(dumpLogFileHandle),
    );

    return transforms;
  }

  static async connect(stream: Duplex) {
    return new this(
      stream,
      await RemoteHost.connect(stream, this.getDumpStreamTap()),
    );
  }
}

export type HostAddedEvent = {
  hostType: HostType;
  host: RemoteHost;
};

export interface Host {
  connect(): Promise<Duplex>;
  displayName: string;
  available: boolean;
  roles: string[];
}

class HostConnections {
  onHostAdded = new SyncEvent<HostAddedEvent>();
  appHost?: HostConnection;
  companionHost?: HostConnection;

  async connect(host: Host, deviceType: DeviceType) {
    const hostTypes: { [key in DeviceType]: HostType } = {
      device: 'appHost',
      phone: 'companionHost',
    };
    const hostType = hostTypes[deviceType];

    const existingHost = this[hostType];
    if (existingHost) existingHost.stream.destroy();

    const stream = await host.connect();

    let hostConnection: HostConnection;
    try {
      hostConnection = await HostConnection.connect(stream);
    } catch (ex) {
      stream.destroy();
      throw ex;
    }

    this[hostType] = hostConnection;
    this.onHostAdded.post({ hostType, host: hostConnection.host });
    return hostConnection;
  }

  async list() {
    const hosts: Host[] = [];

    try {
      for (const device of await developerRelay.hosts()) {
        hosts.push(device);
      }
    } catch (error) {
      throw new Error(
        `An error was encountered when loading the list of available Developer Relay hosts: ${
          (error as Error).message
        }`,
      );
    }

    try {
      for (const device of await USBDebugHost.list()) {
        hosts.push(device);
      }
    } catch (error) {
      throw new Error(
        `An error was encountered when loading the list of available USB hosts: ${
          (error as Error).message
        }`,
      );
    }

    return hosts;
  }

  async listOfType(deviceType: DeviceType) {
    const hosts = await this.list();

    const hostRole = {
      device: 'APP_HOST',
      phone: 'COMPANION_HOST',
    }[deviceType];

    return hosts.filter((host) => host.roles.includes(hostRole));
  }
}

export default HostConnections;
