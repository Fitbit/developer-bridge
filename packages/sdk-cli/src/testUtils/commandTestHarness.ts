import vorpal from '@moleculer/vorpal';

export default function commandTestHarness(extension: (cli: vorpal) => void) {
  const cli = new vorpal();
  // Quiet the MaxListenersExceeded warning when this harness is used in
  // many tests. Each vorpal instance registers a 'vorpal_ui_keypress'
  // listener on the `ui` singleton, and many vorpal instances are
  // created during a normal test run, so the warning is a false alarm.
  cli.ui.setMaxListeners(0);

  cli.use(extension);

  const mockLog = jest.fn();
  const mockPrompt = jest.fn();

  Object.defineProperty(cli, 'activeCommand', {
    value: {
      log: mockLog,
      prompt: mockPrompt,
    },
  });

  const mockUIRedraw: vorpal.Redraw = Object.assign(jest.fn(), {
    clear: jest.fn(),
    done: jest.fn(),
  });

  cli.ui.redraw = mockUIRedraw;
  cli.ui.log = mockLog;
  cli.log = mockLog;

  return {
    cli,
    mockLog,
    mockPrompt,
    mockUIRedraw,
  };
}
