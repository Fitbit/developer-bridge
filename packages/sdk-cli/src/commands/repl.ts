import vorpal from '@moleculer/vorpal';

import HostConnections, { HostConnection } from '../models/HostConnections';

type REPLArgs = vorpal.Args & {
  uuid?: string;
};

export default function repl(stores: { hostConnections: HostConnections }) {
  return (cli: vorpal) => {
    const { hostConnections } = stores;

    const isHostConnected = (host: HostConnection | undefined) => host && !host.host.rpc.ended;

    const exitWithError = (message: string) => {
      cli.activeCommand.log(message);
      (cli as any)._exitMode({
        sessionId: cli.session.id,
      });
    };

    let uuid: string | undefined;

    cli
      .mode('repl device [uuid]')
      .types({ string: ['uuid'] })
      .description('Enter into a REPL with the connected device')
      .delimiter('repl$')
      .init((async (args: REPLArgs) => {
        const hostConnection = stores.hostConnections.appHost;

        if (!isHostConnected(hostConnection)) {
          return exitWithError('No device connected');
        }
        if (!hostConnection!.host.hasEvalSupport()) {
          return exitWithError('Connected device does not support REPL');
        }

        uuid = args.uuid;
        if (uuid) {
          cli.activeCommand.log(`Targeting REPL to UUID: ${uuid}`);
        }

        cli.activeCommand.log('Entering REPL mode, type "exit" to quit');
      }) as any as () => void)
      .action(async (command: string) => {
        const hostConnection = hostConnections.appHost;

        if (!isHostConnected(hostConnection)) {
          cli.activeCommand.log('Host disconnected, exiting REPL');
          cli.execSync('exit');
          return;
        }

        try {
          const result = await hostConnection!.host.eval(command, uuid);
          // No log on failure since a log/trace will be emitted
          if (result.success) cli.activeCommand.log(result.value);
        } catch (ex) {
          cli.activeCommand.log(ex);
        }
      });
  };
}
