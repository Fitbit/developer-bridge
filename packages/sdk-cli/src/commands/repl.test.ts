import vorpal from '@moleculer/vorpal';

import repl from './repl';
import HostConnections from '../models/HostConnections';
import commandTestHarness from '../testUtils/commandTestHarness';

let cli: vorpal;
let mockLog: jest.Mock;
let hostConnections: HostConnections;

beforeEach(() => {
  hostConnections = new HostConnections();
  ({ cli, mockLog } = commandTestHarness(repl({ hostConnections })));
});

const run = (uuid?: string) => uuid ? cli.exec(`repl device ${uuid}`) : cli.exec('repl device');
const cliMode = () => (cli.session as any)._mode;

const mockHost = ({ hasEvalSupport }: { hasEvalSupport: boolean }) => {
  hostConnections.appHost = {
    host: {
      eval: jest.fn(),
      hasEvalSupport: () => hasEvalSupport,
      rpc: {
        ended: false,
      },
    },
  } as any;
};
const mockEval = () => hostConnections.appHost!.host.eval as jest.Mock;

describe('no device is connected', () => {
  beforeEach(() => run());

  it('logs an error', () => {
    expect(mockLog.mock.calls[0][0]).toMatchSnapshot();
  });

  it('exits REPL mode', () => {
    expect(cliMode()).toEqual(false);
  });
});

describe('connected device not support REPL', () => {
  beforeEach(() => {
    mockHost({ hasEvalSupport: false });
    return run();
  });

  it('logs an error', () => {
    expect(mockLog.mock.calls[0][0]).toMatchSnapshot();
  });

  it('exits REPL mode', () => {
    expect(cliMode()).toEqual(false);
  });
});

describe('connected device supports REPL', () => {
  const uuid = 'af93f4e1-de78-4a32-8f39-4b34ee9425df';

  beforeEach(() => mockHost({ hasEvalSupport: true }));

  describe('a UUID is specified', () => {
    beforeEach(() => run(uuid));

    it('sends the statement to the device with UUID', async () => {
      await cli.exec('console');
      expect(hostConnections.appHost!.host.eval).toBeCalledWith('console', uuid);
    });
  });

  describe('no UUID is specified', () => {
    beforeEach(() => run());

    it('logs an informational message explaining how to exit', () => {
      expect(mockLog.mock.calls[0][0]).toMatchSnapshot();
    });

    it('enters REPL mode', () => {
      expect(cliMode()).not.toEqual(false);
    });

    describe('executing a statement', () => {
      beforeEach(() => mockLog.mockClear());

      it('sends the statement to the device', async () => {
        await cli.exec('console');
        expect(hostConnections.appHost!.host.eval).toBeCalledWith('console', undefined);
      });

      it('logs the result on success', async () => {
        const expectedResult = 'is a function';
        mockEval().mockResolvedValueOnce({
          success: true,
          value: expectedResult,
        });
        await cli.exec('console');
        expect(mockLog).toBeCalledWith(expectedResult);
      });

      it('does not log any result on failure', async () => {
        mockEval().mockResolvedValueOnce({
          success: false,
        });
        await cli.exec('console');
        expect(mockLog).not.toBeCalled();
      });

      it('logs if an error occurs during eval', async () => {
        mockEval().mockRejectedValueOnce(
          new Error('something went wrong :('),
        );
        await cli.exec('console');
        expect(mockLog.mock.calls[0][0]).toMatchSnapshot();
      });

      describe('host disconnected since entering REPL', () => {
        beforeEach(() => {
          hostConnections.appHost!.host.rpc.ended = true;
          return cli.exec('console');
        });

        it('logs an error', () => expect(mockLog.mock.calls[0][0]).toMatchSnapshot());
        it('exits the REPL', () => expect(cliMode()).toEqual(false));
      });
    });
  });
});
