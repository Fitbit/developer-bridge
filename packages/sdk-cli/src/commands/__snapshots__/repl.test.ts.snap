// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connected device not support REPL logs an error 1`] = `"Connected device does not support REPL"`;

exports[`connected device supports REPL no UUID is specified executing a statement host disconnected since entering REPL logs an error 1`] = `"Host disconnected, exiting REPL"`;

exports[`connected device supports REPL no UUID is specified executing a statement logs if an error occurs during eval 1`] = `[Error: something went wrong :(]`;

exports[`connected device supports REPL no UUID is specified logs an informational message explaining how to exit 1`] = `"Entering REPL mode, type \\"exit\\" to quit"`;

exports[`no device is connected logs an error 1`] = `"No device connected"`;
